crates/rpc_client/src/lib.rs

```diff
@@ -0,0 +1,1019 @@
+        pub parent_hash: H256,
+        pub sha3_uncles: H256,
+        pub author: Option<Address>,
+        pub state_root: H256,
+        pub transactions_root: H256,
+        pub receipts_root: H256,
+        #[serde(deserialize_with = "optional_u64_from_hex")]
+        pub number: Option<u64>,
+        pub gas_used: U256,
+        pub gas_limit: U256,
+        pub extra_data: Bytes,
+        pub logs_bloom: Option<Bloom>,
+        #[serde(default)]
+        pub timestamp: U256,
+        #[serde(default)]
+        pub difficulty: U256,
+        pub total_difficulty: Option<U256>,
+        #[serde(default, deserialize_with = "deserialize_null_default")]
+        pub seal_fields: Vec<Bytes>,
+        #[serde(default)]
+        pub uncles: Vec<H256>,
+        #[serde(bound = "TX: serde::Serialize + serde::de::DeserializeOwned", default)]
+        pub transactions: Vec<TX>,
+        pub size: Option<U256>,
+        pub mix_hash: Option<H256>,
+        pub nonce: Option<U256>,
+        pub base_fee_per_gas: Option<U256>,
+        pub miner: Address,
+    }
+
+    fn deserialize_null_default<'de, D, T>(deserializer: D) -> Result<T, D::Error>
+    where
+        T: Default + serde::Deserialize<'de>,
+        D: serde::Deserializer<'de>,
+    {
+        use serde::Deserialize;
+        let opt = Option::deserialize(deserializer)?;
+        Ok(opt.unwrap_or_default())
+    }
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum RpcClientError {
+    #[error("Response was not of the expected type")]
+    InterpretationError {
+        msg: String,
+        request_body: String,
+        expected_type: String,
+        response_text: String,
+    },
+
+    #[error("Failed to send request")]
+    SendError { msg: String, request_body: String },
+
+    #[error("Failed to get response body")]
+    ResponseError { msg: String, request_body: String },
+
+    #[error(transparent)]
+    OtherError(#[from] std::io::Error),
+}
+
+pub struct AccountData {
+    pub balance: U256,
+    pub code: Bytes,
+    pub transaction_count: U256,
+}
+
+pub struct RpcClient {
+    url: String,
+    client: reqwest::blocking::Client,
+}
+
+impl RpcClient {
+    pub fn new(url: &str) -> Self {
+        RpcClient {
+            url: url.to_string(),
+            client: reqwest::blocking::Client::new(),
+        }
+    }
+
+    fn make_id() -> Result<u64, std::time::SystemTimeError> {
+        use std::time::{SystemTime, UNIX_EPOCH};
+        let since_epoch = SystemTime::now().duration_since(UNIX_EPOCH)?;
+        Ok(since_epoch
+            .as_secs()
+            .wrapping_sub(since_epoch.subsec_nanos() as u64))
+    }
+
+    pub fn get_tx_by_hash(&self, tx_hash: &H256) -> Result<eth::Transaction, RpcClientError> {
+        use RpcClientError::{InterpretationError, ResponseError, SendError};
+
+        let request_id =
+            jsonrpc::Id::Num(RpcClient::make_id().expect("error generating request ID"));
+
+        let request_body: String = format!(
```

Comment:
Could you reduce code duplication by encoding this in functions+data types? The `EthRequest` type in https://github.com/foundry-rs/foundry/blob/d099e448d5d1603162a6d7a710904b3fba9b7eb5/anvil/core/src/eth/mod.rs
(from issue assigned to you) shows how an example of such a "request type". You should be able to achieve the same using that pattern.

Reply:
Addressed in https://github.com/NomicFoundation/hardhat/pull/3383/commits/9e60ef672a571be81f1dab27b28c60e182a162cd, though perhaps I didn't do it in as fancy a way as you were hoping for, but I did de-duplicate. We can discuss.